core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1941,11 +1941,21 @@ class KafkaApis(val requestChannel: RequestChannel,
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
     val apiVersionRequest = request.body[ApiVersionsRequest]
+    val expectedClientSoftwareNames = config.expectedClientSoftwareNames
+    var hasNonXinfraClient = false
+    if (expectedClientSoftwareNames != null) {
+      expectedClientSoftwareNames.forEach { clientName =>
```

Comment:
I don't understand why we need to loop over expectedClientSoftwareNames, this handleApiVersionRequest is invoked once during client initiation, so this client is either a xinfra client or a kafka client, shouldn't the check be following:

var isCurrentClientXinfra = (!clientName.toLowerCase.contains("xinfra"))

Reply:
As it stands now expectedClientSoftwareNames is a list so to remain true to the data structure i wanted to iterate through it. Are you saying to do something like take the first entry in the list and check that as the client name because if its overwritten that'll be the client name and if not it'll just be a factory name so it won't matter?

Reply:
expectedClientSoftwareNames is a list of valid client names, i.e. it's a list of valid client names from LKC factories/builders and XC factories/builders, and it will be a static/fixed list upon kafka broker startup.

When this handleApiVersionsRequest is invoked, it means a new client has initiated a connection to this broker, and the request contains a clientName/id (apiVersionRequest.context.clientSoftwareName or something?) What we should do is check if the clientSWName in this api version request is from a xinfra client or a LKC client and flag it.